Python: Port SqlInjection

yoff opened the pull request:
Models for the `django` sinks. Probably good to get review at this point.
Should I also include models for `sqlite3`?

There is support for `Model`s, but the subclass detection is a bit approximate...

RasmusWL commented:
Can you add a few words of justification for why we shouldn't copy what they have in JS (`DatabaseAccess`) or what they have in Go (`QueryString`)? (I'm not suggesting it's wrong to do so, but if other languages have a known pattern for this, I think the _default_ action should be to copy from them).

yoff commented:
I ddi not see `QueryString` because they put it in frameworks rather than concepts. It does make sense to have an `SQL` module in concepts and have `QueryString` in there.

I feel that `DatabaseAccess` is an interesting concept, but not the relevant one for sql injection. This because you can access a database in many ways that do not involve sql but could still pose problems or transfer taint.

I understand sql injection to be the execution of user-supplied sql statements. It felt natural to  have a concept of places where such executions can happen and then have a pattern similar to command execution and serialisation, calling out tainted input to such places.

We can do without that concept and instead adopt `QueryString` from Go, which directly point to strings that are being executed. Or we could have both concepts and possibly define one based on the other.

That was probably a few more words than you asked for 😄 thoughts?

RasmusWL commented on python/ql/test/experimental/library-tests/frameworks/django/SqlExecution.py:
```diff
@@ -0,0 +1,31 @@
+    User.objects.raw("SELECT * FROM users WHERE username = %s", (username,))  # $getSql="SELECT * FROM users WHERE username = %s"
+
+    # BAD -- other ways of executing raw SQL code with string interpolation
+    User.objects.annotate(RawSQL("insert into names_file ('name') values ('%s')" % username))  # $getSql=BinaryExpr
+    User.objects.raw("insert into names_file ('name') values ('%s')" % username)  # $getSql=BinaryExpr
+    User.objects.extra("insert into names_file ('name') values ('%s')" % username)  # $getSql=BinaryExpr
```
These methods actually operate on QuerySets, so you can do `User.objects.filter(...).join(...).extra(...)` I don't think the old tests took care to actually illustrate that behavior, but I think we need to take care to do so when porting the tests. (so we're sure on what we _actually_ support)

yoff replied:
Ah, cool, so we need to model these as type-preserving operations.

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,260 @@ private import experimental.dataflow.DataFlow
+          ClassModelSubclass() { node.getId() = any(ClassModelSubclassDef cd).getName() }
+        }
+
+        /** Gets a reference to the `objects` object of a model. */
```
```suggestion
        /** Gets a reference to the `objects` object of a django model. */
```

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,260 @@ private import experimental.dataflow.DataFlow
+          string getName() { result = name }
+        }
+
+        /**
+         * A reference to a class that is a subclass of the `django.db.models.Model` class.
+         * This is quite an approximation, since it simply matches identifiers.
+         */
+        class ClassModelSubclass extends DataFlow::CfgNode {
+          override NameNode node;
+
+          ClassModelSubclass() { node.getId() = any(ClassModelSubclassDef cd).getName() }
+        }
```
Why don't you use a type-tracker for this?

yoff replied:
I did not see how to do that, where do I make it start?

yoff replied:
Oh, I make it start where it is defined now and then track..

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,260 @@ private import experimental.dataflow.DataFlow
-private module Django { }
+/**
+ * Provides models for the `django` PyPI package.
+ * See https://django.palletsprojects.com/en/1.1.x/.
```
I would prefer to use the docstring I made in https://github.com/github/codeql/pull/4490/files#diff-7cbf166da1fec37a36d342fb500f780b1ef87b1d94a828f06e1d0dc616ddfaadR15 :stuck_out_tongue:  -- if you remove this change, we'll automatically get mine after a merge :+1:

yoff replied:
I might also take `django_attr` from there, and only have my one supported case. Then merging will be a bit smoother.

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,260 @@ private import experimental.dataflow.DataFlow
+      /** Gets a reference to the `django.db.connection` object. */
+      DataFlow::Node connection() { result = connection(DataFlow::TypeTracker::end()) }
+
+      /** Gets a reference to the `django.db.connection.cursor` class. */
+      private DataFlow::Node classCursor(DataFlow::TypeTracker t) {
+        t.start() and
+        result = DataFlow::importNode("django.db.connection.cursor")
+        or
+        t.startInAttr("cursor") and
+        result = connection()
+        or
+        exists(DataFlow::TypeTracker t2 | result = classCursor(t2).track(t2, t))
+      }
+
+      /** Gets a reference to the `django.db.connection.cursor` class. */
+      DataFlow::Node classCursor() { result = classCursor(DataFlow::TypeTracker::end()) }
+
+      /** Gets a reference to an instance of `django.db.connection.cursor`. */
+      private DataFlow::Node cursor(DataFlow::TypeTracker t) {
+        t.start() and
+        result.asCfgNode().(CallNode).getFunction() = classCursor().asCfgNode()
+        or
+        exists(DataFlow::TypeTracker t2 | result = cursor(t2).track(t2, t))
+      }
+
+      /** Gets a reference to an instance of `django.db.connection.cursor`. */
+      DataFlow::Node cursor() { result = cursor(DataFlow::TypeTracker::end()) }
```
At this point, I think we're getting into dangerous waters in regards to naming, since there is no `django.db.connection.cursor` class. -- `connection` is defined in [`django/db/__init__.py`](https://github.com/django/django/blob/197b55c53469cf8344d1ba35175236780cb83bd1/django/db/__init__.py#L41), and does some funky forwarding.

I also don't think we've had to track the result of a function class like this before, but hopefully we can get to a nice solution :pray: 

Even though we can't even be sure that the `cursor` attribute is a class (it might just be a proxy function), I think modeling it as though it _is_ a class is very useful. However, I would would suggest we go with naming in the QLDoc along the lines of

```ql
/** Gets a reference to the `cursor` method on `django.db.connection`. */
DataFlow::Node classCursor() { result = classCursor(DataFlow::TypeTracker::end()) }


/** Gets a reference to a result of calling the `cursor` method on `django.db.connection`. */
DataFlow::Node cursor() { result = cursor(DataFlow::TypeTracker::end()) }
```

looking at the code, it's very strange that the `cursor()` predicate is the result of calling the `cursor` method... but I think if we adopt the `cursor::classRef` and `cursor::instance` pattern, it will become more obvious... maybe we should call this `cursor::methodRef` and `cursor::methodResult` instead?

yoff replied:
I see, `connection`, which should perhaps be supplemented by `connections`, ends up giving you a `ConnectionHandler` and `cursor` is just some property on that.

RasmusWL replied:
Also adding support for connections would be great. (found some docs for it [here](https://docs.djangoproject.com/en/3.1/topics/db/multi-db/#using-raw-cursors-with-multiple-databases))

RasmusWL reviewed (changes requested):
Nice :+1:

I think we're going to have merge conflicts with this PR and https://github.com/github/codeql/pull/4490 -- too late now to do anything smart to avoid it ever happening :sweat_smile: :shrug: 

I think the modeling of Django models and QuerySets is a bit tricky. After doing half a review of that part, I would actually propose we leave that out of this PR. I think that with a few minor adjustments, the connection/cursor modeling looks good, and we can merge that (and if you agree, please ignore any comments on that bit for now -- and just keep them in mind for later on).

A nitpick for this whole PR is how you model classes. Would be great if you could adopt the approach we agreed on in https://github.com/github/codeql/pull/4429 with `Foo::classRef` and `Foo::instance` instead of predicates `classFoo` and `foo`. I _do_ need to rewrite some old library code to follow this pattern, so I'm not going to block you on this one, but it would be nice to follow newest pattern in new code :wink:

yoff commented on python/ql/test/experimental/library-tests/frameworks/django/SqlExecution.py:
```diff
@@ -0,0 +1,31 @@
+from django.db.models.expressions import RawSQL
+
+
+def test_plain(username):
+    # GOOD -- Using parameters
+    connection.cursor().execute("SELECT * FROM users WHERE username = %s", username)  # $getSql="SELECT * FROM users WHERE username = %s"
+
+    # BAD -- Using string formatting
+    connection.cursor().execute("SELECT * FROM users WHERE username = '%s'" % username)  # $getSql=BinaryExpr
+
+
+def test_context(username):
+    with connection.cursor() as cursor:
+        # GOOD -- Using parameters
+        cursor.execute("SELECT * FROM users WHERE username = %s", username)  # $getSql="SELECT * FROM users WHERE username = %s"
+
+        # BAD -- Using string formatting
+        cursor.execute("SELECT * FROM users WHERE username = '%s'" % username)  # $getSql=BinaryExpr
```
I agree. I copied the relevant parts of the test file over, but the role here is different, so I should've adapted the code more.

yoff commented on python/ql/test/experimental/library-tests/frameworks/django/SqlExecution.py:
```diff
@@ -0,0 +1,31 @@
+from django.db.models.expressions import RawSQL
+
+
+def test_plain(username):
+    # GOOD -- Using parameters
+    connection.cursor().execute("SELECT * FROM users WHERE username = %s", username)  # $getSql="SELECT * FROM users WHERE username = %s"
+
+    # BAD -- Using string formatting
+    connection.cursor().execute("SELECT * FROM users WHERE username = '%s'" % username)  # $getSql=BinaryExpr
+
+
+def test_context(username):
+    with connection.cursor() as cursor:
+        # GOOD -- Using parameters
+        cursor.execute("SELECT * FROM users WHERE username = %s", username)  # $getSql="SELECT * FROM users WHERE username = %s"
+
+        # BAD -- Using string formatting
+        cursor.execute("SELECT * FROM users WHERE username = '%s'" % username)  # $getSql=BinaryExpr
```
Right, which we probably do not.

yoff commented on python/ql/test/experimental/query-tests/Security-new-dataflow/CWE-089/SqlInjection.expected:
```diff
@@ -0,0 +1,3 @@
+edges
+nodes
+#select
```
Ah, I expected the django sources to be merged by the time this got in, so that we would have results. But if flask is preferred, I will get sources right away :-)

RasmusWL replied:
Preference for `flask` over `django` is certainly debatable, it's mainly just my personal opinion :wink: but I hope you agree? :pray:

yoff commented on python/ql/test/experimental/library-tests/frameworks/django/SqlExecution.py:
```diff
@@ -0,0 +1,31 @@
+    User.objects.raw("SELECT * FROM users WHERE username = %s", (username,))  # $getSql="SELECT * FROM users WHERE username = %s"
+
+    # BAD -- other ways of executing raw SQL code with string interpolation
+    User.objects.annotate(RawSQL("insert into names_file ('name') values ('%s')" % username))  # $getSql=BinaryExpr
+    User.objects.raw("insert into names_file ('name') values ('%s')" % username)  # $getSql=BinaryExpr
+    User.objects.extra("insert into names_file ('name') values ('%s')" % username)  # $getSql=BinaryExpr
```
There seem to be number of ways to pass in arguments, including inside dictionaries. I will try to write a more complete suite. So in general, the first step is to port the existing tests, and the second one is to flesh them out?

RasmusWL replied:
> So in general, the first step is to port the existing tests, and the second one is to flesh them out?

Yes, that is sadly required. I think I wrote up some of the old SQL injection tests, so I'm not entirely without blame in this case :grimacing: :cold_sweat:

yoff commented:
I think I addressed your review now, except I left in the models modelling. I agree that a full model is complicated and should probably include both Models, Managers, and QuerySets as well as QuerySet execution. What is there now, should generally point to real problems, though, which is probably why it was added (piecemeal?) to the existing implementation.
Is it better to cut it now and add something better in the future than leave it now and improve it in the future?

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,313 @@ private import experimental.dataflow.DataFlow
+    override DataFlow::Node getSql() { result.asCfgNode() = node.getArg(0) }
+  }
+
+  /** A call to the `raw` function on a model. */
```
```suggestion
  /** 
   * A call to the `extra` function on a model. 
   *
   * See https://docs.djangoproject.com/en/3.1/ref/models/querysets/#extra
   */
```

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,313 @@ private import experimental.dataflow.DataFlow
+    override DataFlow::Node getSql() { result.asCfgNode() = raw.getArg(0) }
+  }
+
+  /** A call to the `raw` function on a model. */
```
```suggestion
  /** 
   * A call to the `raw` function on a model. 
   *
   * See 
   * - https://docs.djangoproject.com/en/3.1/topics/db/sql/#django.db.models.Manager.raw
   * - https://docs.djangoproject.com/en/3.1/ref/models/querysets/#raw
   */
```

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,313 @@ private import experimental.dataflow.DataFlow
+
+    ObjectsExtra() { node.getFunction() = django::db::models::objects_attr("extra").asCfgNode() }
+
+    override DataFlow::Node getSql() {
+      result.asCfgNode() =
+        [node.getArg([0 .. 5]),
+            node.getArgByName(["select", "where", "params", "tables", "order_by", "select_params"])]
+    }
+  }
```
If you do expand functionality of the parts of the API we talked about not trying to fully support, I think you need to go the full distance and add tests as well :stuck_out_tongue:

Besides this, it is actually safe to pass in user-controlled data to `params` and `select_params`

yoff replied:
I will take those out, although there might be the issue of having quoted placeholders. But since I have done nothing with that anywhere yet, it seems `params` and `select_params` should indeed be ignored for now.

RasmusWL replied:
:+1: to me, `params` and `select_params` are considered safe, but your SQL template can be unsafe it if includes `"where username = '%s'"`, since you shouldn't add the quoting yourself (and we should indeed handle this in the same for `cursor().execute(SQL, params...)`)

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,313 @@ private import experimental.dataflow.DataFlow
+
+    ObjectsAnnotate() {
+      node.getFunction() = django::db::models::objects_attr("annotate").asCfgNode() and
+      raw in [node.getArg(0), node.getArgByName(_)] and
```
Good catch with the "any keyword argument" part. I'm curious if `annotate(RawSQL(foo), RawSQL(bar))` is allowed?

yoff replied:
It should be, the signature is `annotate(*args, **kwargs)`

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,313 @@ private import experimental.dataflow.DataFlow
+        /** Gets a reference to the `django.db.models.expressions` object. */
+        DataFlow::Node expressions() { result = expressions(DataFlow::TypeTracker::end()) }
+
+        /** Gets a reference to the `django.db.models.expressions.RawSQL` class. */
+        private DataFlow::Node classRawSQL(DataFlow::TypeTracker t) {
+          t.start() and
+          result = DataFlow::importNode("django.db.models.expressions.RawSQL")
+          or
+          t.startInAttr("RawSQL") and
+          result = expressions()
+          or
+          exists(DataFlow::TypeTracker t2 | result = classRawSQL(t2).track(t2, t))
+        }
+
+        /** Gets a reference to the `django.db.models.expressions.RawSQL` class. */
+        DataFlow::Node classRawSQL() { result = classRawSQL(DataFlow::TypeTracker::end()) }
```
I think this should be rewritten to use `RawSQL::classRef`

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -7,4 +7,313 @@ private import experimental.dataflow.DataFlow
+    CallNode raw;
+
+    ObjectsAnnotate() {
+      node.getFunction() = django::db::models::objects_attr("annotate").asCfgNode() and
+      raw in [node.getArg(0), node.getArgByName(_)] and
+      raw.getFunction() = django::db::models::classRawSQL().asCfgNode()
```
Does it need to be a requirement that RawSQL is called in-place as an argument to annotate?

Meaning that the following code is invalid

```py
raw = RawSQL(foo)
Users.objects.annotate(val=raw)
```

---

I'm guessing the answer is no, but I would propose we wait with handling this case until later on. If you could add a test-case showing this unhandled pattern, that would be great :+1:

yoff replied:
I was hoping for that to be handled, I will certainly add the test-case.

RasmusWL reviewed (changes requested):
> I agree that a full model is complicated and should probably include both Models, Managers, and QuerySets as well as QuerySet execution. What is there now, should generally point to real problems, though, which is probably why it was added (piecemeal?) to the existing implementation.
> Is it better to cut it now and add something better in the future than leave it now and improve it in the future?

Sure, we can leave it in, but let's make sure we record that the current implementation is not complete, so we don't just forget about it :sweat_smile:


Nitpick:
Would be very useful with inline references for the functions -- not too big a deal for me to google `django raw function`, but just clicking a link is easier :smile: -- I had to look some of them up, and added suggestions for this in a couple of places.

Nitpick:
I would have liked https://github.com/github/codeql/pull/4476/commits/e1dfbc048697d70fea8d3e214c98fc748048e565 to be split into a multiple commits (makes it easier to focus on _one_ thing when reviewing commit by commit):
- one for rewriting `cursor` models
- one for rewriting django Model models
- one for rewriting the test file
- one for new handling of annotate/extra function (they could also be split), which also expands the test-file with the new cases we now support

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -341,8 +341,7 @@ private module Django {
     override DataFlow::Node getSql() {
       result.asCfgNode() =
-        [node.getArg([0 .. 5]),
-            node.getArgByName(["select", "where", "params", "tables", "order_by", "select_params"])]
```
I'm guessing we need to exclude the indexes of the positional arguments corresponding to `select_params` and `params`?

yoff replied:
Good point.

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -307,7 +307,7 @@ private module Django {
     ObjectsAnnotate() {
       node.getFunction() = django::db::models::objects_attr("annotate").asCfgNode() and
-      raw in [node.getArg(0), node.getArgByName(_)] and
+      raw in [node.getArg(_), node.getArgByName(_)] and
```
can we get a test for this? :blush: :pray:

RasmusWL reviewed (changes requested):
nice to see fixes coming in, just spotted a few things.

yoff commented:
Think that is everything. Should be nicely separated into commits. I actually did the indirect `RawSQL` argument even though you told me not to 😊 . Will tackle the merge now.

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -70,6 +70,301 @@ private module Django {
+            DataFlow::Node classRef() { result = classRef(DataFlow::TypeTracker::end()) }
+
+            /** Gets an instance of the `django.db.models.expressions.RawSQL` class. */
+            private DataFlow::Node classInstance(DataFlow::TypeTracker t, ControlFlowNode sql) {
```
```suggestion
            private DataFlow::Node instance(DataFlow::TypeTracker t, ControlFlowNode sql) {
```

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -70,6 +70,301 @@ private module Django {
+            }
+
+            /** Gets an instance of the `django.db.models.expressions.RawSQL` class. */
+            DataFlow::Node classInstance(ControlFlowNode sql) {
+              result = classInstance(DataFlow::TypeTracker::end(), sql)
```
```suggestion
            DataFlow::Node instance(ControlFlowNode sql) {
              result = instance(DataFlow::TypeTracker::end(), sql)
```

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -70,6 +70,301 @@ private module Django {
+                c.getArg(0) = sql
+              )
+              or
+              exists(DataFlow::TypeTracker t2 | result = classInstance(t2, sql).track(t2, t))
```
```suggestion
              exists(DataFlow::TypeTracker t2 | result = instance(t2, sql).track(t2, t))
```

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -70,6 +70,301 @@ private module Django {
+
+      ObjectsAnnotate() {
+        node.getFunction() = django::db::models::objects_attr("annotate").asCfgNode() and
+        django::db::models::expressions::RawSQL::classInstance(sql).asCfgNode() in [node.getArg(_),
```
```suggestion
        django::db::models::expressions::RawSQL::instance(sql).asCfgNode() in [node.getArg(_),
```

RasmusWL commented on python/ql/test/experimental/library-tests/frameworks/django/SqlExecution.py:
```diff
@@ -0,0 +1,28 @@
+
+def test_model():
+    User.objects.raw("some sql")  # $getSql="some sql"
+    User.objects.annotate(RawSQL("some sql"))  # $getSql="some sql"
```
```suggestion
    User.objects.annotate(RawSQL("some sql"))  # $getSql="some sql"
    User.objects.annotate(RawSQL("foo"), RawSQL("bar"))  # $getSql="foo" $getSql="bar"
```

RasmusWL commented on python/ql/src/experimental/semmle/python/frameworks/Django.qll:
```diff
@@ -70,6 +70,301 @@ private module Django {
+      }
+    }
+
+    /**
+     * A call to the `annotate` function on a model using a `RawSQL` argument.
+     *
+     * See https://docs.djangoproject.com/en/3.1/ref/models/querysets/#annotate
+     */
+    private class ObjectsAnnotate extends SqlExecution::Range, DataFlow::CfgNode {
+      override CallNode node;
+      ControlFlowNode sql;
+
+      ObjectsAnnotate() {
+        node.getFunction() = django::db::models::objects_attr("annotate").asCfgNode() and
+        django::db::models::expressions::RawSQL::classInstance(sql).asCfgNode() in [node.getArg(_),
+              node.getArgByName(_)]
+      }
+
+      override DataFlow::Node getSql() { result.asCfgNode() = sql }
+    }
```
I think in general, this is the wrong way to handle taint-propagation, and instead we should make the result of `RawSQL(tainted)`  also be tainted (or maybe even data-flow preserving).

With the current setup, if you have code like this (which might not make any sense)

```py
# foo.py

from sql_helper import sql_helper

def handle_http_request(request):
    raw_sql = RawSQL("select age from user where username = '" + request.GET["username"] + "'")
    sql_result = sql_helper(raw_sql)
    ...

# sql_helper.py

def sql_helper(raw_sql):
    return User.objects.annotate(raw_sql)
```

We would generate an alert that goes from the `request` parameter in `handle_http_request`, to the construction of `RawSQL` -- but it wouldn't be obvious from the alert _where_ the SQL is actually executed :open_mouth: 

Personally I don't know if handling this correctly will ever give us interesting results, so let us just record that this bit could be more polished, and move on to more important things :wink:

yoff replied:
Would we? The sink is the call to `annotate`. Although I do agree on the principle of preferring taint/data-flow over type tracking.

RasmusWL replied:
Well the sink ends up being the result of `getSql` on a `SqlExecution::Range` right?

RasmusWL reviewed (changes requested):
Good stuff :muscle: 

Would like 2 things before approving (both suggestions):

- consistent naming for `instance` predicate
- the extra test-case I asked for

RasmusWL reviewed (approved):
With the last few changes, this LGTM now :1st_place_medal: 

(remember to collect future work identified in our internal tracking issue)
